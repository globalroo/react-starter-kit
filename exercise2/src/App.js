// Conditional rendering example, calling out to endpoint
// Basic GraphQL introduction with server, GraphiQL and adjustable client query

import React, { Component, Fragment } from "react";
import { getQueryFromServer } from "./graphql-service";

// import resources like this to use in your component
// Alternatively you can use a url on the component or <link> via index.html for css as per usual
import logo from "./logo.svg";
import "./App.css";

// Figure out the required data shape in GraphiQL (localhost:4000/graphql)
const getVideosQuery = "{ videos { key, title, synopsis }}";

// TODO: Task - Make this work too
// TODO - Task (bonus points) - Display a list of related Videos to the selected Video
//const getSpecificVideoQuery = "{ video(id:2) { key, title }}";

const myGraphQL = getVideosQuery;

// TODO: Task - Move these to their own files - make them components, add some style.
const Video = ({ video = {} }) => <li>{JSON.stringify(video)}</li>;

const Videos = ({ videos = [] }) => (
	// TODO: Task - Output the movies as a list of poster images with title below
	<ul>
		{videos.map(({ key, ...videoMeta }) => (
			<li key={key}>{JSON.stringify(videoMeta)}</li>
		))}
	</ul>
);

class App extends Component {
	constructor(props) {
		super(props);

		// Changing this initial state via setState triggers a re-render of the component
		this.state = {
			errors: [],
			loaded: false,
			video: null,
			videos: []
		};
	}

	async componentDidMount() {
		// This is the lifecycle method to use to get data from an external service
		// https://reactjs.org/docs/state-and-lifecycle.html
		const result = await getQueryFromServer(myGraphQL);
		// Update state and trigger re-render
		this.setState({ ...result });
	}

	render() {
		// Conditional Rendering example - Conditional return
		if (this.state.errors.length > 0) {
			// TODO: Task - Make Error handling component that looks nice
			return <h3>GraphQL Error: {JSON.stringify(this.state.errors)}</h3>;
		}

		// TODO: Task - Identify and create component optimisations below.
		return (
			<div className="App">
				<header className="App-header">
					<img src={logo} className="App-logo" alt="GraphQL logo" />
					<p>
						Exercise 2 - <code>Using Endpoints</code>
					</p>
				</header>
				<div className="App-main">
					{/* Conditional rendering inline contrived example - Ternary */}
					{this.state.loaded ? (
						<>
							{/* A Fragment allows adjacent / sibling components to be returned without a parent */}
							{/* Conditional rendering inline contrived example - && */}
							{this.state.video !== null && <Video video={this.state.video} />}
							{this.state.videos.length > 0 && (
								<Videos videos={this.state.videos} />
							)}
						</>
					) : (
						<div>Loading...</div>
					)}
				</div>
			</div>
		);
	}
}

export default App;
