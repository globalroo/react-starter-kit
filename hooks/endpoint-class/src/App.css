.App {
	text-align: center;
}

.App-logo {
	animation: App-logo-spin infinite 20s linear;
	height: 10vmin;
}

.App-body {
	background-color: #242525;
	min-height: 100vh;
	display: flex;
	flex-direction: column;
	align-items: center;
	justify-content: center;
	font-size: calc(10px + 2vmin);
	color: white;
}

.App-link {
	color: #61dafb;
}

@keyframes App-logo-spin {
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(360deg);
	}
}

.center-center,
.loader {
	display: -webkit-box;
	display: -moz-box;
	display: -webkit-flex;
	display: -ms-flexbox;
	display: flex;
	-webkit-box-align: center;
	-moz-box-align: center;
	-o-box-align: center;
	-ms-flex-align: center;
	-webkit-align-items: center;
	align-items: center;
	-webkit-box-pack: center;
	-moz-box-pack: center;
	-o-box-pack: center;
	-ms-flex-pack: center;
	-webkit-justify-content: center;
	justify-content: center;
}

.loadingSpinner {
	height: 400px;
}

.App-main {
	font-size: 26px;
	/* font-family: "Press Start 2P"; */
}

.App-main li {
	list-style: none;
}
