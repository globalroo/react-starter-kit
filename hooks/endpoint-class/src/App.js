import React, { Component } from "react";
import { getQueryFromServer } from "./graphql-service";
import logo from "./logo.svg";
import "./App.css";
import { Videos } from "./videos";

const getVideosQuery = "{ videos { key, title }}";

const Loading = () => <p>Loading...</p>;
const Errors = ({ errors }) => <p>{JSON.stringify(errors)}</p>;
export default class App extends Component {
	state = {
		errors: [],
		videos: []
	};

	async componentDidMount() {
		const result = await getQueryFromServer(getVideosQuery);
		this.setState({ ...result });
	}

	render() {
		const { errors, videos } = this.state;
		const hasErrors = errors.length > 0;
		const hasVideos = videos.length > 0 && !hasErrors;
		const isLoading = !hasErrors && !hasVideos;

		return (
			<div className="App">
				<div className="App-body">
					<img src={logo} className="App-logo" alt="GraphQL logo" />
					<p>
						Class - <code>Using Endpoints</code>
					</p>
					<div className="App-main">
						{isLoading && <Loading />}
						{hasVideos && <Videos videos={videos} />}
						{hasErrors && <Errors errors={errors} />}
					</div>
				</div>
			</div>
		);
	}
}
