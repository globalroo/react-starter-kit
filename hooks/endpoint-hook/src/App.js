import React, { useState, useEffect } from "react";
import { getQueryFromServer } from "./graphql-service";
import { Videos } from "./videos";
import logo from "./logo.svg";
import "./App.css";

const getVideosQuery = "{ videos { key, title }}";

const Loading = () => <p>Loading...</p>;
const Errors = ({ errors }) => <p>{JSON.stringify(errors)}</p>;

export default function App() {
	const [response, setResponse] = useState({});

	useEffect(() => {
		getQueryFromServer(getVideosQuery).then(response => {
			setResponse({ ...response });
		});
	}, []);

	const { errors = [], videos = [] } = response;
	const hasErrors = errors.length > 0;
	const hasVideos = videos.length > 0 && !hasErrors;
	const isLoading = !hasErrors && !hasVideos;

	return (
		<div className="App">
			<div className="App-body">
				<img src={logo} className="App-logo" alt="GraphQL logo" />
				<p>
					Hooks - <code>Using Endpoints</code>
				</p>
				<div className="App-main">
					{isLoading && <Loading />}
					{hasVideos && <Videos videos={videos} />}
					{hasErrors && <Errors errors={errors} />}
				</div>
			</div>
		</div>
	);
}
