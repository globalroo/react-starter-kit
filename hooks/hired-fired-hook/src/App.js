import React, { Component, useState } from "react";
import logo from "./logo.svg";
import "./App.css";

import { listData } from "./list-data";

// TODO: Task - optimise components and add to their own files
// TODO: Task - Add hooks to their own files

function StaffList({ data = [] }) {
	const {
		values,
		pushValue,
		popValue,
		updateValueAt,
		removeValueAt
	} = useManagedListHook(data);

	const name = useInputHook("");
	const team = useInputHook("");
	const position = useInputHook("");

	const addPerson = e =>
		pushValue({ name: name.value, team: team.value, position: position.value });

	const removePerson = e => popValue();

	const removePositionFour = e => removeValueAt(4);

	// TODO: Task - This is a hard wired example - can you adjust to update inline?
	// NB: Zero based index
	const updatePositionFour = e =>
		updateValueAt(
			{
				name: name.value,
				team: team.value,
				position: position.value
			},
			4
		);

	// TODO: Task - Make it look nice
	return (
		<div>
			<ul>
				{values.map(item => {
					const { id, ...colleague } = item;
					return <li key={id}>{colleague.name}</li>;
				})}
			</ul>
			<hr />
			<input {...name} placeholder="name" />
			<br />
			<input {...team} placeholder="team" />
			<br />
			<input {...position} placeholder="position" />
			<br />
			<button onClick={addPerson}>Add new colleague</button>
			<button onClick={removePerson}>Fire last person hired</button>
			<button onClick={updatePositionFour}>Update position [4]</button>
			<button onClick={removePositionFour}>Remove position [4]</button>
		</div>
	);
}

class App extends Component {
	render() {
		return (
			<div className="App">
				<header className="App-header">
					<img src={logo} className="App-logo" alt="GraphQL logo" />
					<p>
						Hooks - <code>Using Endpoints</code>
					</p>
					<StaffList data={listData} />
				</header>
			</div>
		);
	}
}

const useInputHook = initialValue => {
	const [value, setValue] = useState(initialValue);
	const onChange = e => setValue(e.target.value);
	return {
		value,
		onChange
	};
};

const useManagedListHook = (initialValue = []) => {
	const [values, setValues] = useState(initialValue);

	const insertValueAt = (item, location) => {
		const addItem = { ...item, id: new Date().getTime() };
		setValues([
			...values.slice(0, location),
			addItem,
			...values.slice(location)
		]);
	};

	const removeValueAt = location =>
		setValues([...values.slice(0, location), ...values.slice(location + 1)]);

	const popValue = () => removeValueAt(values.length - 1);
	const pushValue = item => insertValueAt(item, values.length);

	const updateValueAt = (item, location) => {
		const updatedValues = values.map((listItem, index) => {
			if (index !== location) {
				return listItem;
			}
			return {
				...listItem,
				...item
			};
		});
		setValues(updatedValues);
	};

	return {
		insertValueAt,
		popValue,
		pushValue,
		removeValueAt,
		updateValueAt,
		values
	};
};

export default App;
