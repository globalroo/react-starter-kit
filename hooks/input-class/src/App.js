import React, { Component } from "react";
import "./App.css";

class App extends Component {
	constructor(props) {
		super(props);
		this.state = {
			name: ""
		};
		this.updateName = this.updateName.bind(this);
	}

	updateName(event) {
		this.setState({
			name: event.target.value
		});
	}

	render() {
		return (
			<div className="App">
				<header className="App-header">
					<p>Controlled input using class / state</p>
					<input
						value={this.state.name}
						onChange={this.updateName}
						placeholder="Enter a value"
					/>
					<p>{this.state.name}</p>
				</header>
			</div>
		);
	}
}

export default App;
