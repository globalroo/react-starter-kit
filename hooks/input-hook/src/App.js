import React, { useState } from "react";
import "./App.css";

export default function App() {
	const [name, setName] = useState("");
	const updateName = e => setName(e.target.value);
	return (
		<div className="App">
			<header className="App-header">
				<p>Controlled input using hooks</p>
				<input value={name} onChange={updateName} placeholder="Enter a value" />
				<p>{name}</p>
			</header>
		</div>
	);
}
