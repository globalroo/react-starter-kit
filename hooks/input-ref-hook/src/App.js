import React, {
	Component,
	createRef,
	useRef
	// useImperativeMethods,
	// forwardRef
} from "react";
import logo from "./logo.svg";
import "./App.css";

const BasicRefHook = () => {
	const inputRef = useRef(); //Initialise a ref hook to set on DOM
	const onButtonClick = () => alert(inputRef.current.value);
	return (
		<>
			<input ref={inputRef} type="text" />
			<button onClick={onButtonClick}>Pull value from DOM</button>
		</>
	);
};

//WiP
// let BindRefAndUseInParent = ({ myRef }) => {
// 	const inputRef = useRef(myRef);
// 	useImperativeMethods(myRef, () => ({
// 		focus: () => {
// 			inputRef.current.style.color = "red";
// 			inputRef.current.focus();
// 		}
// 	}));
// 	return <input ref={inputRef} placeholder="Press and check console" />;
// };
// BindRefAndUseInParent = forwardRef(BindRefAndUseInParent);

class App extends Component {
	myRef = createRef();
	test = () => {
		// console.log(this.myRef.current); // We have a handle on child DOMNode now.
		// this.myRef.current.focus();
	};
	render() {
		return (
			<div className="App">
				<header className="App-header">
					<img src={logo} className="App-logo" alt="logo" />
					<p>Input ref example</p>
					<BasicRefHook />
					<br />
					{/* I want to access a child dom node from this class, pass the ref to the child so it can send it back (and augment it the Node with extra magic!) */}
					{/* <BindRefAndUseInParent myRef={this.myRef} /> */}
					<button onClick={this.test}>Do some action</button>
				</header>
			</div>
		);
	}
}

export default App;
