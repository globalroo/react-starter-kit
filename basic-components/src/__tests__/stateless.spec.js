import React from "react";
import renderer from "react-test-renderer";

import { render, fireEvent } from "react-testing-library";
import App from "src/App";
import {
	Basic,
	BasicWithProps,
	BasicWithDOMAttrs,
	BasicWithEventHandler,
	BasicWithEventSentToParent,
	BasicWithValuePassedToParent,
	BasicWithChildren
} from "src/stateless";

// Using Kent C Dodds testing library
it("Finds some text in the page", () => {
	const { getByText } = render(<App />);
	expect(getByText("Learn React")).toBeInTheDocument();
});

it("Clicks a button and calls back to a method", () => {
	const spy = jest.fn();
	const { getByText } = render(
		<BasicWithEventSentToParent handleClick={spy} />
	);
	fireEvent.click(getByText("Click Me"));
	expect(spy).toHaveBeenCalledTimes(1);
});

it("Clicks a button and receives a 'Called' payload", () => {
	const spy = jest.fn();
	const { getByText } = render(
		<BasicWithValuePassedToParent handleClick={spy} />
	);
	fireEvent.click(getByText("Click Me"));
	expect(spy).toHaveBeenCalledWith("Clicked");
});

it("Clicks a button and receives a 'Called' payload", () => {
	const { getByText } = render(
		<BasicWithChildren>
			<p>Hello World</p>
		</BasicWithChildren>
	);
	expect(getByText("Hello World")).toBeInTheDocument();
});

// Standard React patterns, React-test-renderer - different stuff
// Wide ranging catch all pattern to just flag any changes to your components
describe("Simple snapshot tests for simple components using basic react-renderer", () => {
	test("I can render a <Basic /> component", () => {
		const component = renderer.create(<Basic />).toJSON();
		expect(component).toMatchSnapshot();
	});
	test("I can render a <BasicWithProps /> component", () => {
		const component = renderer
			.create(<BasicWithProps message={"Test"} />)
			.toJSON();
		expect(component).toMatchSnapshot();
	});
	test("I can render a <BasicWithDOMAttrs /> component", () => {
		const component = renderer.create(<BasicWithDOMAttrs />).toJSON();
		expect(component).toMatchSnapshot();
	});
	test("I can render a <BasicWithEventHandler /> component", () => {
		const component = renderer.create(<BasicWithEventHandler />).toJSON();
		expect(component).toMatchSnapshot();
	});
});
