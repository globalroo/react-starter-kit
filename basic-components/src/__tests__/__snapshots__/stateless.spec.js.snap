// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple snapshot tests for simple components using basic react-renderer I can render a <Basic /> component 1`] = `
<div>
  Basic Component
</div>
`;

exports[`Simple snapshot tests for simple components using basic react-renderer I can render a <BasicWithDOMAttrs /> component 1`] = `
<div>
  <input
    className="my-class"
    data-qa="qa-hook"
    readOnly={true}
    tabIndex="1"
  />
</div>
`;

exports[`Simple snapshot tests for simple components using basic react-renderer I can render a <BasicWithEventHandler /> component 1`] = `
<div>
  <button
    onClick={[Function]}
  >
    Click Me
  </button>
</div>
`;

exports[`Simple snapshot tests for simple components using basic react-renderer I can render a <BasicWithProps /> component 1`] = `
<div>
  Test
</div>
`;
